---
title: ""  
author: "<img src='https://github.com/RajLabMSSM/`r read.dcf('DESCRIPTION', fields = 'Package')[1]`/raw/`r gsub('[*] ','',system('git branch', intern = TRUE)[1])`/inst/hex/hex.png' height='300'><br><br>
        `r badger::badge_github_version(color = 'black')` 
        `r badger::badge_github_actions(action = 'R-CMD-check-bioc')`
        `r badger::badge_last_commit()`
        `r badger::badge_codecov()` 
        `r badger::badge_license()` 
        <h5>Author: <i>Brian M. Schilder</i></h5>" 
date: "<h5>README updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h5>"
output:
  github_document
---

```{r, echo=FALSE, include=FALSE}
pkg <- read.dcf("DESCRIPTION", fields = "Package")[1]
description <- read.dcf("DESCRIPTION", fields = "Description")[1]
```
 
## ``r pkg``: `r gsub("echoverse module: ","", description)`

This R package is part of the *echoverse* suite that supports [`echolocatoR`](https://github.com/RajLabMSSM/echolocatoR):
an automated genomic fine-mapping pipeline. 

If you use ``r pkg``, please cite: 

> `r citation(pkg)$textVersion`

and

> N Kerimov, JD Hayhurst, K Peikova, et al. A compendium of uniformly 
processed human gene expression and splicing quantitative trait loci. 
Nat Genet 53, 1290â€“1299 (2021). [https://doi.org/10.1038/s41588-021-00924-w](https://doi.org/10.1038/s41588-021-00924-w)


## Installation 

### [1] *Optional step: conda*   

To ensure all dependencies are installed and don't conflict with each other, 
you can create a *conda* environment by downloading [this *yaml* file](https://github.com/RajLabMSSM/catalogueR/blob/master/inst/conda/catalogueR.yml) and 
entering the following in command line:  
```
conda env create -n <path_to_yml>/catalogueR.yml
```

### [2] Install `catalogueR` in R  

#### Tabix 

If you don't use the *conda* env, you will also need to make sure [tabix](http://www.htslib.org/doc/tabix.html) is installed.

**NOTE**: Different distributions of *tabix* may work with 
varying degrees of consistency. 
I recommend you try them in the following order of preference:

1. [brew](https://formulae.brew.sh/formula/htslib) (for Mac users).
2. [source](http://www.htslib.org/download/) 
3. [bioconda](https://anaconda.org/bioconda/tabix) 
  - As of Feb 2021, I noticed that the bioconda distribution of *tabix* 
  will randomly fail at queries. It appears this hasn't been updated 
  in over 2 years, so I do NOT recommend you use it.

```R
if(!require("remotes")) install.packages("remotes")

remotes::install_github("RajLabMSSM/`r pkg`")
library(`r pkg`)
``` 

## Documentation 

### [Website](https://rajlabmssm.github.io/`r pkg`) 
### [Getting started](https://rajlabmssm.github.io/`r pkg`/articles/`r pkg`) 


## Notes

- The functions in `catalogueR` are partly derived from the 
[*eQTL Catalogue* tutorial](http://htmlpreview.github.io/?https://github.com/eQTL-Catalogue/eQTL-Catalogue-resources/blob/master/scripts/eQTL_API_usecase.html).  
- [The ALT allele is always the effect allele in eQTL Catalogue](https://www.ebi.ac.uk/eqtl/Data_access/). 

- Additional *eQTL Catalogue* resources:  
  + [GitHub repository](https://github.com/eQTL-Catalogue/eQTL-Catalogue-resources)  
  + [In-depth API documentation](https://www.ebi.ac.uk/eqtl/api-docs/) 
  + FTP server: *ftp://ftp.ebi.ac.uk/pub/databases/spot/eQTL/csv* 
 
<hr>

## Contact

<a href="https://bschilder.github.io/BMSchilder/" target="_blank">Brian
M. Schilder, Bioinformatician II</a>  
<a href="https://rajlab.org" target="_blank">Raj Lab</a>  
<a href="https://icahn.mssm.edu/about/departments/neuroscience" target="_blank">Department
of Neuroscience, Icahn School of Medicine at Mount Sinai</a>  

