---
title: "`catalogueR`: query" 
author: "Brian M. Schilder"
date: "Most recent update:<br> `r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{query}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE, error = TRUE} 
root.dir <- "~/Desktop"
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  root.dir = root.dir,
  fig.height = 12,
  fig.width = 10
)  
knitr::opts_knit$set(root.dir = root.dir)
# knitr::opts_chunk$get("root.dir")
```


# Tutorial: Query


```{r setup}
library(catalogueR) 
```


## Introduction  
 
*eQTL Catalogue* includes a large number of standardized QTL datasets (110 datasets from 20 studies as of  7/4/2020). It actually contains more than just eQTL data. For each dataset, the following kinds of QTLs can be queried:

- **gene expression QTL**: `quant_method="ge"` (*default*) or `quant_method="microarray"`, depending on the dataset. catalogueR will automatically select whichever option is available.  
- **exon expression QTL**: *under construction* `quant_method="ge"`  
- **transcript usage QTL** *under construction* `quant_method="tx"`  
- **promoter, splice junction and 3สน end usage QTL** *under construction* `quant_method="txrev"`   


## Metadata  

You can view of a metadata table for all current datasets:  

```{r}
data("meta")

createDT(meta)
```

You can seach through the metadata for datasets with certain keywords (case-insensitive substrings across multiple columns). 

```{r}
qtl_datasets <- eQTL_Catalogue.search_metadata(qtl_search=c("Alasoo_2018","monocyte"))

print(qtl_datasets)
```



### [Approach 1] Query with summary stats  

Supply one or more paths to [GWAS] summary stats files (one per locus) and automatically download any eQTL data within that range. The files can be any of these formats, either *gzip*-compressed (`.gz`) or uncompressed: `.csv`, `.tsv`, `space-separated`  

<br>

The summary stats files must have the following column names (order doesn't matter):
  - `SNP` (rsid for each SNP)
  - `CHR` (chromosome; with or without the "chr" prefix is fine)
  - `POS` (basepair position)
  - ... (optional extra columns)

```{r [Approach 1] Query with summary stats } 
sumstats_paths <- example_sumstats_paths()

gwas.qtl_paths <- eQTL_Catalogue.query(sumstats_paths = sumstats_paths["BST1"],  
                                       qtl_search = c("myeloid","Alasoo_2018"),
                                       split_files  = TRUE,    
                                       merge_with_gwas  = TRUE,
                                       force_new_subset  = TRUE,
                                       progress_bar  = FALSE,
                                       nThread = 1)
```

- Because you selected the argument `split_files=T` the query results have been distributed across multiple files and saved to disk (useful when you don't want to load one massive data.table every time you want to look at specific results subsets).  
- In this scenario, `eQTL_Catalogue.query()` returns a list of those file paths (instead of the data.table itself).  
- To merge these files back together and import them into R, use the following function:  

```{r}
GWAS.QTL <- gather_files(file_paths = gwas.qtl_paths)
# Interactive datatable of results 
## WARNING: Don't use this function on large datatables, might cause freezing.
createDT(head(GWAS.QTL))
```



### [Approach 2] Query with coordinates  

You can also makes queries to *eQTL Catalogue* by manually specifying the coordinates of the region you want to extract,
as well as the `unique_id` of the QTL dataset (see `data("meta")` for IDs).
 
```{r [Approach 2] Query with coordinates  } 
GWAS.QTL_manual <- eQTL_Catalogue.fetch(unique_id="Alasoo_2018.macrophage_IFNg",  
                                        nThread = 4,  
                                        chrom = 8, 
                                        bp_lower=21527069,
                                        bp_upper=23525543)

createDT(head(GWAS.QTL_manual))
```


# Session info 

<details> 

```{r Session info}
utils::sessionInfo()
```

</details>
