<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>QTL data extraction and colocalization • catalogueR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="QTL data extraction and colocalization">
<meta property="og:description" content="API access to eQTL Catalogue full summary stats and colocalization functions.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">catalogueR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/intro.html">catalogueR: intro</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RajLabMSSM/catalogueR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="cataloguer" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#cataloguer" class="anchor"></a>catalogueR</h1></div>
<p>Rapid querying, colocalization, and plotting of summary stats from the eQTL Catalogue.</p>
<div id="intro" class="section level2">
<h2 class="hasAnchor">
<a href="#intro" class="anchor"></a>Intro</h2>
<p>The functions in <strong>catalogueR</strong> are partly derived from the following <a href="http://htmlpreview.github.io/?https://github.com/eQTL-Catalogue/eQTL-Catalogue-resources/blob/master/scripts/eQTL_API_usecase.html">eQTL Catalogue tutorial</a>.<br>
Additional eQTL Catalogue Resources:</p>
<ul>
<li>
<a href="https://github.com/eQTL-Catalogue/eQTL-Catalogue-resources">GitHub</a><br>
</li>
<li>
<a href="https://www.ebi.ac.uk/eqtl/api-docs/">In-depth API documentation</a><br>
</li>
<li><a href="ftp://ftp.ebi.ac.uk/pub/databases/spot/eQTL/csv">FTP server</a></li>
</ul>
<p><em>NOTE</em>: <a href="https://www.ebi.ac.uk/eqtl/Data_access/">The ALT allele is always the effect allele in eQTL Catalogue</a>.</p>
<p>Notes on parallelization: There’s multiple levels to parallelize on. You can only chooose one level at a time:</p>
<ul>
<li>
<code>multithread_qtl=T</code>: Across QTL datasets<br>
</li>
<li>
<code>multithread_loci=T</code>: Across loci<br>
</li>
<li>
<code>multithread_tabix=T</code>: Within tabix files</li>
</ul>
<p>You can also get a speedup by using tabix instead of the rest API Test: For 3 loci, and X QTL datasets:</p>
<ul>
<li>
<strong>RESTful API</strong>: 7.5 minutes<br>
</li>
<li>
<strong>Tabix</strong>: 27 seconds (*clear winner! ~17x speedup)<br>
That said, if you’re only querying a small number of specific SNPs (rather than a large range), the RESTful API can sometimes be faster.</li>
</ul>
<hr>
</div>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h2>
<div id="installation" class="section level3">
<h3 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h3>
<div class="sourceCode" id="cb1"><pre class="r"><span class="kw">if</span>(!<span class="st">"devtools"</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span>()){<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>)}
<span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"RajLabMSSM/catalogueR"</span>)</pre></div>
</div>
<div id="command-line-dependencies" class="section level3">
<h3 class="hasAnchor">
<a href="#command-line-dependencies" class="anchor"></a>Command line dependencies</h3>
<ul>
<li>
<a href="http://www.htslib.org/doc/tabix.html">tabix</a> (Install via <a href="https://anaconda.org/bioconda/tabix">conda</a> or from <a href="http://www.htslib.org/download/">source</a>)</li>
</ul>
</div>
<div id="intro-example" class="section level3">
<h3 class="hasAnchor">
<a href="#intro-example" class="anchor"></a>Intro example</h3>
<div class="sourceCode" id="cb2"><pre class="r"><span class="co"># Returns both the gwas_data you supplied and the queried QTL summary stats, </span>
<span class="co">## all merged into one data.table.  </span>

<span class="no">sumstats_paths</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/example_sumstats_paths.html">example_sumstats_paths</a></span>()

<span class="no">gwas.qtl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/eQTL_Catalogue.query.html">eQTL_Catalogue.query</a></span>(<span class="co"># Any number of summary stats files</span>
                           <span class="kw">sumstats_paths</span> <span class="kw">=</span> <span class="no">sumstats_paths</span>,

                           <span class="co"># Name of each locus (a vector of the same length as the sumstats_paths)  </span>
                           <span class="co"># If none are supplied (NULL), names will be assigned based on the chromosomal coordinates.</span>
                           <span class="kw">loci_names</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"BST1"</span>,<span class="st">"LRRK2"</span>,<span class="st">"MEX3C"</span>),

                           <span class="co"># The file you want the merged gwas/qtl results to be saved in.</span>
                           <span class="co">## The file will be tab-delimited, but you can choose to leave it uncompressed by removing the '.gz' at the end of the path here.</span>
                           <span class="kw">output_path</span><span class="kw">=</span><span class="st">"./example_data/Nalls23andMe_2019/eQTL_Catalogue.tsv.gz"</span>,

                           <span class="co"># This function will automatically search for any datasets that match your criterion.</span>
                           <span class="co">## For example, if you search "Alasoo_2018", it will query the datasets</span>
                           <span class="co">##"Alasoo_2018.macrophage_naive","Alasoo_2018.macrophage_IFNg",Alasoo_2018.macrophage_Salmonella", </span>
                           <span class="co">## and "Alasoo_2018.macrophage_IFNg+Salmonella").</span>
                           <span class="co">## You can be more specific about which datasets you want to include, for example by searching: "Alasoo_2018.macrophage_IFNg".</span>
                           <span class="co">## You can even search by tissue or condition type (e.g.c("blood","brain")) </span>
                           <span class="co">##and any QTL datasets containing those terms will be quried too.</span>
                           <span class="kw">qtl_search</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ROSMAP"</span>,<span class="st">"Alasoo_2018"</span>),

                           <span class="co"># Tabix is about ~17x faster (default; =T) than the REST API (=F).</span>
                           <span class="kw">use_tabix</span><span class="kw">=</span><span class="no">T</span>,

                           <span class="co"># Use multiple CPU cores to speed up your queries</span>
                           <span class="kw">nThread</span><span class="kw">=</span><span class="fl">4</span>,

                           <span class="co"># Multi-thread across QTL datasets (good when you're querying lots of QTL datasets.)</span>
                           <span class="kw">multithread_qtl</span><span class="kw">=</span><span class="no">T</span>,

                           <span class="co"># Multi-thread across loci (good when you have lots of gwas loci)</span>
                           <span class="kw">multithread_loci</span><span class="kw">=</span><span class="no">F</span>,

                           <span class="co"># Save the results as one file per QTL dataset (with all loci within each file).</span>
                           <span class="co">## If this is set to `=T`, then this function will return the list of paths where these files were saved.</span>
                           <span class="co">### A helper function is provided to import and merge them back together in R.</span>
                           <span class="co">## If this is set to `=F`, then this function will instead return one big merged data.table</span>
                           <span class="co">### containing results from all QTL datasets and all loci. </span>
                           <span class="co">### `=F` is not recommended when you have many large loci and/or many QTL datasets, </span>
                           <span class="co">### because you can only fit so much data into memory.</span>
                           <span class="kw">split_files</span><span class="kw">=</span><span class="no">T</span>)</pre></div>
<hr>
</div>
</div>
<div id="author" class="section level2">
<h2 class="hasAnchor">
<a href="#author" class="anchor"></a>Author</h2>
<p>Brian M. Schilder<br>
Raj Lab<br>
Icahn School of Medicine at Mount Sinai<br>
New York, New York, USA</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/RajLabMSSM/catalogueR/">https://​github.com/​RajLabMSSM/​catalogueR/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/RajLabMSSM/catalogueR/issues">https://​github.com/​RajLabMSSM/​catalogueR/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Brian Schilder <br><small class="roles"> Author, maintainer </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Brian Schilder.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
