<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>catalogueR: colocalize • catalogueR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="catalogueR: colocalize">
<meta property="og:description" content="catalogueR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">catalogueR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/colocalize.html">catalogueR: colocalize</a>
    </li>
    <li>
      <a href="../articles/intro.html">catalogueR: query</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RajLabMSSM/catalogueR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="colocalize_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>catalogueR: colocalize</h1>
                        <h4 class="author">Brian M. Schilder</h4>
            
            <h4 class="date">Most recent update:<br> 2021-02-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/RajLabMSSM/catalogueR/blob/master/vignettes/colocalize.Rmd"><code>vignettes/colocalize.Rmd</code></a></small>
      <div class="hidden name"><code>colocalize.Rmd</code></div>

    </div>

    
    
<div id="tutorial-colocalize" class="section level1">
<h1 class="hasAnchor">
<a href="#tutorial-colocalize" class="anchor"></a>Tutorial: Colocalize</h1>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RajLabMSSM/catalogueR">catalogueR</a></span><span class="op">)</span> 
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span></code></pre></div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<ul>
<li><p>Here, we aim to more robustly test whether the genetic signals underlying two dataset (e.g. GWAS vs. eQTL in the same locus)are indeed the same, using a methodology called colocalization.</p></li>
<li><p>Specifically, we will use <a href="https://github.com/chr1swallace/coloc"><code>coloc</code></a>, which infers the probability that each SNP is causal in a given locus in each of the datasets. It then tests the hypothesis that those signals show substantially similar association distributions.</p></li>
</ul>
</div>
<div id="list-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#list-datasets" class="anchor"></a>List datasets</h2>
<ul>
<li><p>We have previously queried <em>eQTL Catalogue</em> using several Parkinson’s disease GWAS loci (with <code><a href="../reference/eQTL_Catalogue.query.html">eQTL_Catalogue.query()</a></code>).</p></li>
<li><p>Let’s first gather those files and save them to disk as csv files.</p></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">query_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"BST1"</span><span class="op">=</span><span class="fu">catalogueR</span><span class="fu">::</span><span class="va"><a href="../reference/BST1__Alasoo_2018.macrophage_IFNg.html">BST1__Alasoo_2018.macrophage_IFNg</a></span>,
                  <span class="st">"LRRK2"</span><span class="op">=</span><span class="fu">catalogueR</span><span class="fu">::</span><span class="va"><a href="../reference/LRRK2__Alasoo_2018.macrophage_IFNg.html">LRRK2__Alasoo_2018.macrophage_IFNg</a></span>,
                   <span class="st">"MEX3C"</span><span class="op">=</span><span class="fu">catalogueR</span><span class="fu">::</span><span class="va"><a href="../reference/MEX3C__Alasoo_2018.macrophage_IFNg.html">MEX3C__Alasoo_2018.macrophage_IFNg</a></span>
<span class="op">)</span>

<span class="co"># select where you want to store them </span>
<span class="va">storage_dir</span> <span class="op">&lt;-</span> <span class="st">"catalogueR_queries/Nalls23andMe_2019"</span>

<span class="va">gwas.qtl_paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">query_res</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span>
  <span class="va">storage_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">storage_dir</span>,
                            <span class="st">"Alasoo_2018.macrophage_IFNg"</span>,
                            <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">x</span>,<span class="st">"__Alasoo_2018.macrophage_IFNg.tsv.gz"</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">dirname</a></span><span class="op">(</span><span class="va">storage_path</span><span class="op">)</span>, showWarnings <span class="op">=</span> <span class="cn">F</span>, recursive <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
  <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fwrite.html">fwrite</a></span><span class="op">(</span><span class="va">query_res</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>, <span class="va">storage_path</span><span class="op">)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">storage_path</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span> <span class="op">%&gt;%</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">`names&lt;-`</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">query_res</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="run-coloc" class="section level1">
<h1 class="hasAnchor">
<a href="#run-coloc" class="anchor"></a>Run coloc</h1>
<ul>
<li>Since its original release, <a href="https://github.com/chr1swallace/coloc"><code>coloc</code></a> has been updated so that it can now model multiple causal variants within a given dataset (see the <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1008720">paper</a>), whereas previously it could assume one causal variants. Thus, it may be able to better estimate the colocalization probability between two datasets.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coloc_QTLs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_coloc.html">run_coloc</a></span><span class="op">(</span><span class="va">gwas.qtl_paths</span>,
                        save_path<span class="op">=</span><span class="st">"./coloc_results.tsv.gz"</span>,
                        nThread<span class="op">=</span><span class="fl">4</span>,
                        top_snp_only<span class="op">=</span><span class="cn">T</span>,
                        split_by_group<span class="op">=</span><span class="cn">F</span>,
                        method <span class="op">=</span> <span class="st">"abf"</span><span class="op">)</span>
<span class="co">#&gt; + QTL Group = Alasoo_2018.macrophage_IFNg</span></code></pre></div>
<div id="plot-results" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-results" class="anchor"></a>Plot results</h2>
<p>First, let’s plot only the results with &gt;80% colocalization probability. This is a colocalization threshold commonly used in the field.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coloc_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_coloc_summary.html">plot_coloc_summary</a></span><span class="op">(</span>coloc_QTLs <span class="op">=</span> <span class="va">coloc_QTLs</span>,
                                 PP_thresh <span class="op">=</span> <span class="fl">.8</span><span class="op">)</span>
<span class="co">#&gt; Error in plot_coloc_summary(coloc_QTLs = coloc_QTLs, PP_thresh = 0.8): </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ** No coloc tests reached the PPH4 threshold (0.8).Try lowering `PP_thresh`. **</span></code></pre></div>
<ul>
<li><p>Uh oh! Looks like there were no colocalizations at 80% probability. Let’s explore the data a bit and lower the threshold.</p></li>
<li><p>Indeed, two loci show mild colocalization at 50% probability. As a general rule, the more tests you do to higher your probability threshold should be, since Bayesian methods don’t calculate P-values (and this cannot used multiple-testing correction methods like FDR).</p></li>
<li><p>For example, if you ran many thousands of coloc test across many GWAS-eQTL pairs, it’s advisable to raise your PP threshold to .90 or even .99. This also has the benefit of narrowing down your (sometimes many!) results so that you may focus on only the strongest colocalizations.</p></li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coloc_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_coloc_summary.html">plot_coloc_summary</a></span><span class="op">(</span>coloc_QTLs <span class="op">=</span> <span class="va">coloc_QTLs</span>,
                                 PP_thresh <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span></code></pre></div>
<p><img src="colocalize_files/figure-html/Plot%20results%202-1.png" width="960"></p>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<details><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.3 (2020-10-10)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Big Sur 10.16</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] dplyr_1.0.4      catalogueR_0.1.0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] highr_0.8         compiler_4.0.3    pillar_1.4.7      tools_4.0.3      </span>
<span class="co">#&gt;  [5] digest_0.6.27     evaluate_0.14     memoise_2.0.0     lifecycle_1.0.0  </span>
<span class="co">#&gt;  [9] tibble_3.0.6      gtable_0.3.0      pkgconfig_2.0.3   rlang_0.4.10     </span>
<span class="co">#&gt; [13] DBI_1.1.1         parallel_4.0.3    yaml_2.2.1        pkgdown_1.6.1    </span>
<span class="co">#&gt; [17] xfun_0.21         fastmap_1.1.0     stringr_1.4.0     knitr_1.31       </span>
<span class="co">#&gt; [21] generics_0.1.0    desc_1.2.0        fs_1.5.0          vctrs_0.3.6      </span>
<span class="co">#&gt; [25] systemfonts_1.0.1 rprojroot_2.0.2   grid_4.0.3        tidyselect_1.1.0 </span>
<span class="co">#&gt; [29] glue_1.4.2        data.table_1.13.6 R6_2.5.0          textshaping_0.3.0</span>
<span class="co">#&gt; [33] rmarkdown_2.6     farver_2.0.3      tidyr_1.1.2       purrr_0.3.4      </span>
<span class="co">#&gt; [37] ggplot2_3.3.3     magrittr_2.0.1    scales_1.1.1      htmltools_0.5.1.1</span>
<span class="co">#&gt; [41] ellipsis_0.3.1    assertthat_0.2.1  colorspace_2.0-0  labeling_0.4.2   </span>
<span class="co">#&gt; [45] ragg_1.1.0        stringi_1.5.3     munsell_0.5.0     cachem_1.0.3     </span>
<span class="co">#&gt; [49] crayon_1.4.1</span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Brian Schilder.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
