<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Iterate queries to <em>eQTL Catalogue</em> — eQTL_Catalogue.query • catalogueR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Iterate queries to <em>eQTL Catalogue</em> — eQTL_Catalogue.query" />
<meta property="og:description" content="Deetermines which datasets to query using qtl_search.
Uses coordinates from stored summary stats files (e.g. GWAS) 
to determine which regions to query from eQTL Catalogue. 
Each locus file can be stored separately, 
or merged together to form one large file with all query results." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">catalogueR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/intro.html">catalogueR: intro</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/RajLabMSSM/catalogueR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Iterate queries to <em>eQTL Catalogue</em></h1>
    <small class="dont-index">Source: <a href='https://github.com/RajLabMSSM/catalogueR/blob/master/R/query.R'><code>R/query.R</code></a></small>
    <div class="hidden name"><code>eQTL_Catalogue.query.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Deetermines which datasets to query using <code>qtl_search</code>.
Uses coordinates from stored summary stats files (e.g. GWAS) 
to determine which regions to query from <em>eQTL Catalogue</em>. 
Each locus file can be stored separately, 
or merged together to form one large file with all query results.</p>
    </div>

    <pre class="usage"><span class='fu'>eQTL_Catalogue.query</span>(
  <span class='kw'>sumstats_paths</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>output_dir</span> <span class='kw'>=</span> <span class='st'>"./catalogueR_queries"</span>,
  <span class='kw'>qtl_search</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>use_tabix</span> <span class='kw'>=</span> <span class='no'>T</span>,
  <span class='kw'>nThread</span> <span class='kw'>=</span> <span class='fl'>4</span>,
  <span class='kw'>quant_method</span> <span class='kw'>=</span> <span class='st'>"ge"</span>,
  <span class='kw'>infer_region</span> <span class='kw'>=</span> <span class='no'>T</span>,
  <span class='kw'>split_files</span> <span class='kw'>=</span> <span class='no'>T</span>,
  <span class='kw'>merge_with_gwas</span> <span class='kw'>=</span> <span class='no'>T</span>,
  <span class='kw'>force_new_subset</span> <span class='kw'>=</span> <span class='no'>F</span>,
  <span class='kw'>genome_build</span> <span class='kw'>=</span> <span class='st'>"hg19"</span>,
  <span class='kw'>progress_bar</span> <span class='kw'>=</span> <span class='no'>T</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='no'>T</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>sumstats_paths</th>
      <td><p>A list of paths to any number of summary stats files 
whose coordinates you want to use to make queries to eQTL Catalogue.  
If you wish to add custom names to the loci, simply add these as the names of the path list
 (e.g. <code><a href='https://rdrr.io/r/base/c.html'>c(BST1="&lt;path&gt;/&lt;to&gt;/&lt;BST1_file&gt;", LRRK2="&lt;path&gt;/&lt;to&gt;/&lt;LRRK2_file&gt;")</a></code>). 
 Otherwise, loci will automatically named based on their min/max genomic coordinates.</p> 
<p>The minimum columns in these files required to make queries include:</p><dl'>
<dt>SNP</dt><dd><p>RSID of each SNP.</p></dd> 
<dt>CHR</dt><dd><p>Chromosome (can be in "chr12" or "12" format).</p></dd>  
<dt>POS</dt><dd><p>Genomic position of each SNP.</p></dd>  
<dt>...</dt><dd><p>Optional extra columns.</p></dd>  

</dl></td>
    </tr>
    <tr>
      <th>output_dir</th>
      <td><p>The folder you want the merged gwas/qtl results to be saved to 
(set <code>output_dir=F</code> if you don't want to save the results).  
If <code>split_files=F</code>, all query results will be merged into one and saved as <em>&lt;output_dir&gt;/eQTL_Catalogue.tsv.gz</em>.  
If <code>split_files=T</code>, all query results will instead be split into smaller files and stored in <em>&lt;output_dir&gt;/</em>.</p></td>
    </tr>
    <tr>
      <th>qtl_search</th>
      <td><p>This function will automatically search for any datasets that match your criterion.
For example, if you search "Alasoo_2018", it will query the datasets:</p><ul>
<li><p>Alasoo_2018.macrophage_naive</p></li>
<li><p>Alasoo_2018.macrophage_Salmonella</p></li>
<li><p>Alasoo_2018.macrophage_IFNg+Salmonella</p></li>
</ul><p>You can be more specific about which datasets you want to include, 
for example by searching: "Alasoo_2018.macrophage_IFNg".
You can even search by tissue or condition type (e.g.c("blood","brain")) 
and any QTL datasets containing those substrings (case-insensitive) in their name or metadata will be queried too.</p></td>
    </tr>
    <tr>
      <th>use_tabix</th>
      <td><p>Tabix is about ~17x faster (<em>default:</em> =T) than the REST API (<em>=F</em>).</p></td>
    </tr>
    <tr>
      <th>nThread</th>
      <td><p>The number of CPU cores you want to use to speed up your queries through parallelization.</p></td>
    </tr>
    <tr>
      <th>quant_method</th>
      <td><p>eQTL Catalogue actually contains more than just eQTL data.
For each dataset, the following kinds of QTLs can be queried:</p><dl'>
<dt>gene expression QTL</dt><dd><p><code>quant_method="ge"</code> (<em>default</em>) or <code>quant_method="microarray"</code>, depending on the dataset. <strong>catalogueR</strong> will automatically select whichever option is available.</p></dd>  
<dt>exon expression QTL</dt><dd><p><em>*under construction*</em>  <code>quant_method="ex"</code></p></dd>
<dt>transcript usage QTL</dt><dd><p><em>*under construction*</em>  <code>quant_method="tx"</code></p></dd>
<dt>promoter, splice junction and 3' end usage QTL</dt><dd><p><em>*under construction*</em>  <code>quant_method="txrev"</code></p></dd>   

</dl></td>
    </tr>
    <tr>
      <th>split_files</th>
      <td><p>Save the results as one file per QTL dataset (with all loci within each file).
If this is set to <code>=T</code>, then this function will return the list of paths where these files were saved.
A helper function is provided to import and merge them back together in R.  
If this is set to  <code>=F</code>, then this function will instead return one big merged data.table  
containing results from all QTL datasets and all loci.  
<code>=F</code> is not recommended when you have many large loci and/or many QTL datasets,  
because you can only fit so much data into memory.</p></td>
    </tr>
    <tr>
      <th>merge_with_gwas</th>
      <td><p>Whether you want to merge your QTL query results with your GWAS data 
(convenient, but takes up more storage).</p></td>
    </tr>
    <tr>
      <th>force_new_subset</th>
      <td><p>By default, <strong>catalogueR</strong> will use any pre-existing files that match your query. 
Set <code>force_new_subset=T</code> to override this and force a new query.</p></td>
    </tr>
    <tr>
      <th>genome_build</th>
      <td><p>The genome build of your query coordinates (e.g. <code>gwas_data</code>). 
If your coordinates are in <em>hg19</em>, <strong>catalogueR</strong> will automatically lift them over 
to <em>hg38</em> (as this is the build that eQTL Catalogue uses).</p></td>
    </tr>
    <tr>
      <th>progress_bar</th>
      <td><p><code>progress_bar=T</code> allows progress to be monitored even when multithreading enabled.
Requires R package <code>pbmcapply</code>.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Show more (<code>=T</code>) or fewer (<code>=F</code>) messages.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other eQTL Catalogue: 
<code><a href='eQTL_Catalogue.fetch.html'>eQTL_Catalogue.fetch</a>()</code>,
<code><a href='eQTL_Catalogue.iterate_fetch.html'>eQTL_Catalogue.iterate_fetch</a>()</code>,
<code><a href='eQTL_Catalogue.search_metadata.html'>eQTL_Catalogue.search_metadata</a>()</code>,
<code><a href='fetch_restAPI.html'>fetch_restAPI</a>()</code>,
<code><a href='fetch_tabix.html'>fetch_tabix</a>()</code>,
<code><a href='merge_gwas_qtl.html'>merge_gwas_qtl</a>()</code>,
<code><a href='meta.html'>meta</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>sumstats_paths</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='example_sumstats_paths.html'>example_sumstats_paths</a></span>()

<span class='co'># Merged results</span>
<span class='no'>GWAS.QTL</span> <span class='kw'>&lt;-</span> <span class='fu'>eQTL_Catalogue.query</span>(<span class='kw'>sumstats_paths</span><span class='kw'>=</span><span class='no'>sumstats_paths</span>, <span class='kw'>qtl_search</span><span class='kw'>=</span><span class='st'>"Alasoo_2018"</span>, <span class='kw'>nThread</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>force_new_subset</span><span class='kw'>=</span><span class='no'>T</span>, <span class='kw'>merge_with_gwas</span><span class='kw'>=</span><span class='no'>F</span>, <span class='kw'>progress_bar</span><span class='kw'>=</span><span class='no'>T</span>, <span class='kw'>split_files</span><span class='kw'>=</span><span class='no'>F</span>)</div><div class='output co'>#&gt; [1] "+ Optimizing multi-threading..."
#&gt; [1] "++ Multi-threading across QTL datasets."
#&gt; [1] "eQTL_Catalogue:: Querying 4 QTL datasets x 3 GWAS loci (12 total)"</div><div class='output co'>#&gt; <span class='message'>Alasoo_2018.macrophage_naive</span></div><div class='output co'>#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: BST1 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:25:44</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:25:48</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 4 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 78481 SNPs returned in 18.9 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: LRRK2 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:26:07</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:26:13</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 6 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 2795 SNPs returned in 5.8 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: MEX3C ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:26:19</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:26:24</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 5 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 50088 SNPs returned in 9.4 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "+ Returning merged data.table of query results."</div><div class='output co'>#&gt; <span class='message'>Alasoo_2018.macrophage_IFNg</span></div><div class='output co'>#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: BST1 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:26:33</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:26:37</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 4 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 57036 SNPs returned in 11.2 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: LRRK2 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:26:48</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:26:53</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 5 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 3354 SNPs returned in 6.2 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: MEX3C ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:26:59</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:27:02</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 3 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 51536 SNPs returned in 9.7 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "+ Returning merged data.table of query results."</div><div class='output co'>#&gt; <span class='message'>Alasoo_2018.macrophage_Salmonella</span></div><div class='output co'>#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: BST1 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:27:12</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:27:17</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 5 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 68136 SNPs returned in 21.7 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: LRRK2 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:27:38</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:27:42</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 4 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 2795 SNPs returned in 5.4 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: MEX3C ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:27:48</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:27:53</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 5 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 51536 SNPs returned in 9.2 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "+ Returning merged data.table of query results."</div><div class='output co'>#&gt; <span class='message'>Alasoo_2018.macrophage_IFNg+Salmonella</span></div><div class='output co'>#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: BST1 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:28:02</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:28:08</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 6 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 43983 SNPs returned in 14.8 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: LRRK2 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:28:23</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:28:28</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 5 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 3354 SNPs returned in 5.6 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: MEX3C ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:28:33</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:28:37</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 4 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 45995 SNPs returned in 7.4 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "+ Returning merged data.table of query results."
#&gt; [1] "++ Post-processing merged results."
#&gt; [1] "+ Checking results for missing queries..."
#&gt; [1] "+ Saving merged query results ==&gt; ./catalogueR_queries/eQTL_Catalogue.tsv.gz"
#&gt; Time difference of 3.1 mins</div><div class='input'><span class='co'># Merged results (parallel)</span>
<span class='no'>GWAS.QTL</span> <span class='kw'>&lt;-</span> <span class='fu'>eQTL_Catalogue.query</span>(<span class='kw'>sumstats_paths</span><span class='kw'>=</span><span class='no'>sumstats_paths</span>, <span class='kw'>qtl_search</span><span class='kw'>=</span><span class='st'>"Alasoo_2018"</span>, <span class='kw'>nThread</span><span class='kw'>=</span><span class='fl'>4</span>, <span class='kw'>force_new_subset</span><span class='kw'>=</span><span class='no'>T</span>, <span class='kw'>merge_with_gwas</span><span class='kw'>=</span><span class='no'>F</span>, <span class='kw'>progress_bar</span><span class='kw'>=</span><span class='no'>T</span>, <span class='kw'>split_files</span><span class='kw'>=</span><span class='no'>F</span>)</div><div class='output co'>#&gt; [1] "+ Optimizing multi-threading..."
#&gt; [1] "++ Multi-threading across QTL datasets."
#&gt; [1] "eQTL_Catalogue:: Querying 4 QTL datasets x 3 GWAS loci (12 total)"
#&gt; [1] "++ Post-processing merged results."
#&gt; [1] "+ Checking results for missing queries..."
#&gt; [1] "+ QTL datasets with no hits/failed to pull data:"
#&gt; [1] "  + BST1__Alasoo_2018.macrophage_naive"
#&gt; [1] "  + LRRK2__Alasoo_2018.macrophage_naive"
#&gt; [1] "  + MEX3C__Alasoo_2018.macrophage_naive"
#&gt; [1] "  + BST1__Alasoo_2018.macrophage_IFNg"
#&gt; [1] "  + LRRK2__Alasoo_2018.macrophage_IFNg"
#&gt; [1] "  + MEX3C__Alasoo_2018.macrophage_IFNg"
#&gt; [1] "  + BST1__Alasoo_2018.macrophage_Salmonella"
#&gt; [1] "  + LRRK2__Alasoo_2018.macrophage_Salmonella"
#&gt; [1] "  + MEX3C__Alasoo_2018.macrophage_Salmonella"
#&gt; [1] "+ Saving merged query results ==&gt; ./catalogueR_queries/eQTL_Catalogue.tsv.gz"
#&gt; Time difference of 57.5 secs</div><div class='input'>
<span class='co'># Split results</span>
<span class='no'>gwas.qtl</span> <span class='kw'>&lt;-</span> <span class='fu'>eQTL_Catalogue.query</span>(<span class='kw'>sumstats_paths</span><span class='kw'>=</span><span class='no'>sumstats_paths</span>, <span class='kw'>qtl_search</span><span class='kw'>=</span><span class='st'>"Alasoo_2018"</span>, <span class='kw'>nThread</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>force_new_subset</span><span class='kw'>=</span><span class='no'>T</span>, <span class='kw'>merge_with_gwas</span><span class='kw'>=</span><span class='no'>F</span>, <span class='kw'>progress_bar</span><span class='kw'>=</span><span class='no'>T</span>)</div><div class='output co'>#&gt; [1] "+ Optimizing multi-threading..."
#&gt; [1] "++ Multi-threading across QTL datasets."
#&gt; [1] "eQTL_Catalogue:: Querying 4 QTL datasets x 3 GWAS loci (12 total)"</div><div class='output co'>#&gt; <span class='message'>Alasoo_2018.macrophage_naive</span></div><div class='output co'>#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: BST1 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:29:45</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:29:48</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 3 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 78481 SNPs returned in 18.3 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Saving split file ==&gt; ./catalogueR_queries/Alasoo_2018.macrophage_naive/BST1__Alasoo_2018.macrophage_naive.tsv.gz"
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: LRRK2 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:30:06</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:30:11</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 5 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 2795 SNPs returned in 5.8 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Saving split file ==&gt; ./catalogueR_queries/Alasoo_2018.macrophage_naive/LRRK2__Alasoo_2018.macrophage_naive.tsv.gz"
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: MEX3C ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:30:17</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:30:20</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 3 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 50088 SNPs returned in 8.8 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Saving split file ==&gt; ./catalogueR_queries/Alasoo_2018.macrophage_naive/MEX3C__Alasoo_2018.macrophage_naive.tsv.gz"
#&gt; [1] "+ Returning list of split query results files."</div><div class='output co'>#&gt; <span class='message'>Alasoo_2018.macrophage_IFNg</span></div><div class='output co'>#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: BST1 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:30:29</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:30:34</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 5 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 57036 SNPs returned in 16.1 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Saving split file ==&gt; ./catalogueR_queries/Alasoo_2018.macrophage_IFNg/BST1__Alasoo_2018.macrophage_IFNg.tsv.gz"
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: LRRK2 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:30:50</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:30:54</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 4 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 3354 SNPs returned in 5.8 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Saving split file ==&gt; ./catalogueR_queries/Alasoo_2018.macrophage_IFNg/LRRK2__Alasoo_2018.macrophage_IFNg.tsv.gz"
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: MEX3C ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:31:00</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:31:03</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 3 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 51536 SNPs returned in 12.1 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Saving split file ==&gt; ./catalogueR_queries/Alasoo_2018.macrophage_IFNg/MEX3C__Alasoo_2018.macrophage_IFNg.tsv.gz"
#&gt; [1] "+ Returning list of split query results files."</div><div class='output co'>#&gt; <span class='message'>Alasoo_2018.macrophage_Salmonella</span></div><div class='output co'>#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: BST1 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:31:16</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:31:18</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 2 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 68136 SNPs returned in 14.9 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Saving split file ==&gt; ./catalogueR_queries/Alasoo_2018.macrophage_Salmonella/BST1__Alasoo_2018.macrophage_Salmonella.tsv.gz"
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: LRRK2 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:31:34</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:31:38</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 4 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 2795 SNPs returned in 6.1 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Saving split file ==&gt; ./catalogueR_queries/Alasoo_2018.macrophage_Salmonella/LRRK2__Alasoo_2018.macrophage_Salmonella.tsv.gz"
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: MEX3C ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:31:44</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:31:48</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 4 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 51536 SNPs returned in 6.9 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Saving split file ==&gt; ./catalogueR_queries/Alasoo_2018.macrophage_Salmonella/MEX3C__Alasoo_2018.macrophage_Salmonella.tsv.gz"
#&gt; [1] "+ Returning list of split query results files."</div><div class='output co'>#&gt; <span class='message'>Alasoo_2018.macrophage_IFNg+Salmonella</span></div><div class='output co'>#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: BST1 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:31:55</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:32:00</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 5 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 43983 SNPs returned in 16.3 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Saving split file ==&gt; ./catalogueR_queries/Alasoo_2018.macrophage_IFNg+Salmonella/BST1__Alasoo_2018.macrophage_IFNg+Salmonella.tsv.gz"
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: LRRK2 ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:32:16</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:32:20</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 4 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 3354 SNPs returned in 5.8 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Saving split file ==&gt; ./catalogueR_queries/Alasoo_2018.macrophage_IFNg+Salmonella/LRRK2__Alasoo_2018.macrophage_IFNg+Salmonella.tsv.gz"
#&gt; [1] "++ Extracting locus name from `sumstats_paths` names."</div><div class='output co'>#&gt; <span class='message'>_+_+_+_+_+_+_+_+_--- Locus: MEX3C ---_+_+_+_+_+_+_+_+_</span></div><div class='output co'>#&gt; <span class='message'>Start at 2020-07-05 17:32:26</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>End at 2020-07-05 17:32:29</span></div><div class='output co'>#&gt; <span class='message'>Runtime in total is: 3 secs</span></div><div class='output co'>#&gt; [1] "++ Inferring coordinates from gwas_data"
#&gt; [1] "eQTL_Catalogue:: 45995 SNPs returned in 7.4 seconds."
#&gt; [1] "++ Converting: Ensembl IDs ==&gt; HGNC gene symbols"
#&gt; [1] "++ Adding `Locus.GWAS` column."
#&gt; [1] "++ Saving split file ==&gt; ./catalogueR_queries/Alasoo_2018.macrophage_IFNg+Salmonella/MEX3C__Alasoo_2018.macrophage_IFNg+Salmonella.tsv.gz"
#&gt; [1] "+ Returning list of split query results files."
#&gt; [1] "++ Returning list of split files paths."
#&gt; Time difference of 2.9 mins</div><div class='input'><span class='no'>GWAS.QTL</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='gather_files.html'>gather_files</a></span>(<span class='kw'>file_paths</span> <span class='kw'>=</span> <span class='no'>gwas.qtl</span>)</div><div class='output co'>#&gt; [1] "+ Merging 12 files."
#&gt; [1] "+ Using 4 cores."
#&gt; [1] "+ Merged data.table: 459089 rows x 21 columns."</div><div class='input'><span class='co'># Split results (parallel)</span>
<span class='no'>gwas.qtl</span> <span class='kw'>&lt;-</span> <span class='fu'>eQTL_Catalogue.query</span>(<span class='kw'>sumstats_paths</span><span class='kw'>=</span><span class='no'>sumstats_paths</span>, <span class='kw'>qtl_search</span><span class='kw'>=</span><span class='st'>"Alasoo_2018"</span>, <span class='kw'>nThread</span><span class='kw'>=</span><span class='fl'>4</span>, <span class='kw'>force_new_subset</span><span class='kw'>=</span><span class='no'>T</span>, <span class='kw'>merge_with_gwas</span><span class='kw'>=</span><span class='no'>F</span>, <span class='kw'>progress_bar</span><span class='kw'>=</span><span class='no'>T</span>)</div><div class='output co'>#&gt; [1] "+ Optimizing multi-threading..."
#&gt; [1] "++ Multi-threading across QTL datasets."
#&gt; [1] "eQTL_Catalogue:: Querying 4 QTL datasets x 3 GWAS loci (12 total)"
#&gt; [1] "++ Returning list of split files paths."
#&gt; Time difference of 51.3 secs</div><div class='input'><span class='no'>GWAS.QTL</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='gather_files.html'>gather_files</a></span>(<span class='kw'>file_paths</span> <span class='kw'>=</span> <span class='no'>gwas.qtl</span>)</div><div class='output co'>#&gt; [1] "+ Merging 3 files."
#&gt; [1] "+ Using 4 cores."
#&gt; [1] "+ Merged data.table: 131364 rows x 21 columns."</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Brian Schilder.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


